debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java

```diff
@@ -199,19 +195,28 @@ protected void doEnqueue(T record) throws InterruptedException {
         if (LOGGER.isDebugEnabled()) {
             LOGGER.debug("Enqueuing source record '{}'", record);
         }
-        // Waiting for queue to add more record.
-        while (maxQueueSizeInBytes > 0 && currentQueueSizeInBytes.get() > maxQueueSizeInBytes) {
-            Thread.sleep(pollInterval.toMillis());
-        }
-        // If we pass a positiveLong max.queue.size.in.bytes to enable handling queue size in bytes feature
-        if (maxQueueSizeInBytes > 0) {
-            long messageSize = ObjectSizeCalculator.getObjectSize(record);
-            objectMap.put(record, messageSize);
-            currentQueueSizeInBytes.addAndGet(messageSize);
-        }
 
-        // this will also raise an InterruptedException if the thread is interrupted while waiting for space in the queue
-        queue.put(record);
+        synchronized (this) {
+            while (queue.size() >= maxQueueSize || (maxQueueSizeInBytes > 0 && currentQueueSizeInBytes >= maxQueueSizeInBytes)) {
+                // notify poll() to drain queue
+                this.notifyAll();
```

Comment:
`notifyAll` is not recommended as it awaits all threads while only one can read the data.